### DISCUSSION TITLE
mariadb mysql table/database photoprism.photos is corrupt. How do I fix it?

### DISCUSSION THREAD

**Discussion by alpha754293** [photoprism/photoprism#4656] [Q&A]
For some strange reason, my mariadb decided to bomb out earlier today.

In trying to google the issue, I have been able to at least get the mariadb container started with the `--innodb-force-recovery=2` option. (`--innodb-force-recovery=1` still results in the database getting the fatal 11 error.)

I ran the `mariadb-check --all-databases` inside the container (had to add the `--skip-grant-table` option to the `command` line in my `docker-compose.yml` file otherwise, it wouldn't let the root user inside the container, check the database/tables.

After doing that, it says:

```
photoprism.photos
Warning: InnoDB: Index `idx_photos_checked_at` contains 1832435 entries, should be 1832246.
error: corrupt
```

I then tried to run `mariadb-check -r --all-databases` and it said:

(for all of the `photoprism.*` tables)

`note: The storage engine for the table doesn't support repair`

I would prefer to *not* have to rebuild the entire table/database from scratch if I can avoid it, but if it is needed as a last resort, I can do that.

(cf. https://mariadb.com/kb/en/innodb-recovery-modes/)

Also, per that link that I just pasted above, it said that I can dump the database/tables, but I would imagine that I probably wouldn't want to dump a database that is known to be corrupt, where I assume that if I were to dump it and then import it back in, that I will just be importing said known corrupted table/database back into the system, so it won't really fix anything.

If you have any suggestions and/or advice as to how I can fix this, your help would be greatly appreciated.

Thank you.

*edit*
I also just tried this as well:

Entered the `photoprism_mariadb_1` container by running `sudo docker exec -it photoprism_mariadb_1 /bin/bash`

Once I am inside, I ran:

`mariadb mysql -u root`

Once I am inside the MariaDB [mysql] prompt, I ran this;

`OPTIMIZE TABLE photoprism.photos;`

And this is the result of that:

```
+-------------------+----------+----------+-------------------------------------------------------------------+
| Table             | Op       | Msg_type | Msg_text                                                          |
+-------------------+----------+----------+-------------------------------------------------------------------+
| photoprism.photos | optimize | note     | Table does not support optimize, doing recreate + analyze instead |
| photoprism.photos | optimize | error    | Index for table 'photos' is corrupt; try to repair it             |
| photoprism.photos | optimize | status   | Operation failed                                                  |
+-------------------+----------+----------+-------------------------------------------------------------------+
```

At this point, I don't know how I can fix the corrupted index for the photos table anymore.

If you have any suggestions/recommendations, your help is greatly appreciated.

(Sidenote: No, the database isn't mounting via NFS nor SMB. It's on a local disk. Thanks.)

**Comment by dingedan** [comment]
I have the same issue and unfortunately my oldest backup has the corruption as well. Did you manage to resolve this? 

btw, my error for your latest command is different:
`+-------------------+----------+----------+-------------------------------------------------------------------+
| Table             | Op       | Msg_type | Msg_text                                                          |
+-------------------+----------+----------+-------------------------------------------------------------------+
| photoprism.photos | optimize | note     | Table does not support optimize, doing recreate + analyze instead |
| photoprism.photos | optimize | error    | Got error 1000 "No error information" from storage engine InnoDB  |
| photoprism.photos | optimize | status   | Operation failed                                                  |
+-------------------+----------+----------+-------------------------------------------------------------------+`

A CHECK TABLE gives me following:
`
MariaDB [mysql]> CHECK TABLE photoprism.photos;
+-------------------+-------+----------+-------------------------------------------------------------------------+
| Table             | Op    | Msg_type | Msg_text                                                                |
+-------------------+-------+----------+-------------------------------------------------------------------------+
| photoprism.photos | check | Warning  | InnoDB: The B-tree of index PRIMARY is corrupted.                       |
| photoprism.photos | check | Warning  | InnoDB: The B-tree of index uix_photos_photo_uid is corrupted.          |
| photoprism.photos | check | Warning  | InnoDB: The B-tree of index idx_photos_path_name is corrupted.          |
| photoprism.photos | check | Warning  | InnoDB: The B-tree of index idx_photos_photo_lng is corrupted.          |
| photoprism.photos | check | Warning  | InnoDB: The B-tree of index idx_photos_photo_lat is corrupted.          |
| photoprism.photos | check | Warning  | InnoDB: The B-tree of index idx_photos_ymd is corrupted.                |
| photoprism.photos | check | Warning  | InnoDB: The B-tree of index idx_photos_camera_lens is corrupted.        |
| photoprism.photos | check | Warning  | InnoDB: The B-tree of index idx_photos_checked_at is corrupted.         |
| photoprism.photos | check | Warning  | InnoDB: The B-tree of index idx_photos_taken_uid is corrupted.          |
| photoprism.photos | check | Warning  | InnoDB: The B-tree of index idx_photos_place_id is corrupted.           |
| photoprism.photos | check | Warning  | InnoDB: The B-tree of index idx_photos_cell_id is corrupted.            |
| photoprism.photos | check | Warning  | InnoDB: The B-tree of index idx_photos_published_at is corrupted.       |
| photoprism.photos | check | Warning  | InnoDB: The B-tree of index idx_photos_deleted_at is corrupted.         |
| photoprism.photos | check | Warning  | InnoDB: The B-tree of index idx_photos_uuid is corrupted.               |
| photoprism.photos | check | Warning  | InnoDB: The B-tree of index idx_photos_country_year_month is corrupted. |
| photoprism.photos | check | Warning  | InnoDB: The B-tree of index idx_photos_created_by is corrupted.         |
| photoprism.photos | check | error    | Corrupt                                                                 |
+-------------------+-------+----------+-------------------------------------------------------------------------+
`


And when I try to query the photos table the connection closes. All other tables work fine.

**Reply by alpha754293** [reply]
@dingedan 
No. Being that it was for Photoprism, I tried to rebuild and/or recover the database by exporting the database from MySQL itself (putting it into `--innodb-force-recovery=5`) so I was able to read the table(s), tried to export it, and then tried to re-import it back in, and it failed to do both. Then when I tried to put `--innodb-force-recovery=1`, and start Photoprism again, it failed to start completely, so I just deleted the mysql/database folder (I forget what it's called), but whatever the path is -- I just completedly deleted that entire folder, and then forced it to re-index the entire database from scratch.

So, the down side with this solution is that it takes a while and it takes processing power to rebuild the database from scratch, in its entirety, but at least it's like starting off from a clean slate, all over again.

After much googling, I didn't find a solution that actually worked to recover the database, but being that it's for Photoprism, having it rebuild the entire database from scratch again isn't too big of a deal, other than time and processing power required to complete this task.

**Reply by dingedan** [reply]
thanks @alpha754293 : I tried those methods to no avail as well. For me having to rebuild is a bit of a bigger deal because I've spent many many hours entering extra meta data for old scanned photos. For next time I am just making sure to keep my backups for longer and getting a Photoprism specific backup will make things easier than system level backup as well.

**Reply by alpha754293** [reply]
@dingedan 
Yeah - I just needed Photoprism to help manage and index my stuff, so I don't bother with adding extra meta data or anything like that.

The one thing that I've also done is that if the system/database is important, I actually have two Linux Containers that both run Photoprism, indexing from the same source directories, but write out the database to two separate locations so that it's like having a "mirrored" copy (of sorts).

Takes more storage and processing power to do it twice, but this provides some data/evidence that sometimes, even backing up the database/folder is not enough to be able to restore/recover from a corrupted database, etc. (and that attempts at trying to recover from said corrupted database aren't always successful).

**Reply by dingedan** [reply]
Interesting! After deleting the database and re-scanning my images it turns out that all of the metadata that I entered manually is still there. I don't believe Photoprism changes exif data so I had assumed this was all stored in the DB but it must be done somewhere else.

**Reply by graciousgrey** [reply]
In case there is no database backup, metadata can also be restored from YAML backup files: https://docs.photoprism.app/user-guide/backups/export/

**Reply by alpha754293** [reply]
@dingedan 
Admittedly, it's not the *best* method, but in a pinch, it can/does work.

(But for your case, where you are adding additional metadata, it might be a terrible idea as you'd basically have to add the metadata to your pictures twice.)

@graciousgrey 
Thank you for that.

That's actually really good to know.

Pardon my stupid question though -- from the link that you provided, it says that it will also backup the metadata to YAML. How do you import them back in to Photoprism? I couldn't find any place, in the documentation, where the answer to my question might be obvious, so if can please point me to where I might be able to find the answer to that -- that would be greatly appreciated.

**Reply by graciousgrey** [reply]
If yaml files are present in the storage folder during the first indexing of a fresh instance, metadata can be restored from them.

https://docs.photoprism.app/user-guide/backups/restore/
